### PR TITLE
Implement huffman encoding in ocaml

### PR DESCRIPTION
**Pull request by shubhamkumar13**
References -> https://ocaml.org/learn/tutorials/99problems.html

### CODE REVIEW

**Comment by shubhamkumar13** [comment]
huffman code

This might be the one

**Comment by leios** [comment]
Just checked, the codebook generated by this tree is also totally valid!

**Comment by shubhamkumar13** [comment]
Is there something wrong with the implementation? It was closed if I correctly remember

**Comment by leios** [comment]
I think this PR is ready for review; however, I don't have ocaml expertise, so if someone else could step in that would be great!

**Comment by ntindle** [comment]
[lang: ocaml]

**Comment by Amaras** [comment]
Hum... now that I look at it, you need to insert the code in the chapter (markdown) file before we can merge @shubhamkumar13
I don't really have expertise in OCaml, sadly, so I can't review this PR properly.

**Comment by leios** [comment]
If the code is alright, we can probably add the line numbers in and merge this soon? I wasn't sure if the comment from @Amaras meant the code looked good or not

File: contents/huffman_encoding/code/ocaml/huffman.ml [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+  the string that is input
+  For eg. in a string : "Chennai" I want to remove char : 'n'
+  such that we get the old value and the new value as a tuple
+
+  TL;DR : "Chennai" -> ("Cheai", "Chennai")
+*)
+let char_diff str ch = 
+  (String.split_on_char ch str, str) |> fun (new_str, old_str) -> 
+  (List.rev new_str, old_str) |> fun (new_str, old_str) -> 
+  (List.fold_left (fun acc x -> x ^ acc) "" new_str, old_str)
```
Comment (ShadowMitia) [inline contents/huffman_encoding/code/ocaml/huffman.ml:12]: You can simplify `char_diff` with (should be a bit more efficient):

```suggestion
let char_diff str ch = 
  let res = String.concat "" (String.split_on_char ch str) in
  (res, str)
```

File: contents/huffman_encoding/code/ocaml/huffman.ml [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+(* Converts "ABCD" -> ['A'; 'B'; 'C'; 'D'] *)
+let str_to_charlist str = 
+  let lst = ref [] in
+  String.iter (fun x -> lst := x :: !lst) str |> fun _ -> 
+  List.rev !lst
+
```
Comment (ShadowMitia) [inline contents/huffman_encoding/code/ocaml/huffman.ml:19]: You can simplifyt to:
```suggestion
let str_to_charlist s = List.init (String.length s) (String.get s)

```

File: contents/huffman_encoding/code/ocaml/huffman.ml [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+let huffman huffman_tree = prefixes_of_tree "" huffman_tree
+
+(* Helper functions *)
+let char_to_str = Printf.sprintf "%c"
+
+let str_list msg = 
+  List.map char_to_str @@ str_to_charlist msg
+
+let list_to_string lst =
+  List.fold_left (fun acc x -> acc ^ x) "" lst
```
Comment (ShadowMitia) [inline contents/huffman_encoding/code/ocaml/huffman.ml:95]: I believe you can replace the function with this
```suggestion
  String.concat "" lst
```

Calling '^' creates a new string at each iteration. `concat` should be more efficient here.

File: contents/huffman_encoding/code/ocaml/huffman.ml [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+(* Encoding and decoding functions *)
+let encode codebook x =
+  List.filter (fun (ch, code) -> ch = x) codebook |> fun x ->
+  List.hd x |> snd
+
+let encode_msg codebook msg =
+  List.map (fun x -> encode codebook x) @@ str_list msg |>
+  List.fold_left (fun acc x -> acc ^ x) ""
```
Comment (ShadowMitia) [inline contents/huffman_encoding/code/ocaml/huffman.ml:104]: You can reuse `list_to_string` here

```suggestion
   list_to_string (List.map (fun x -> encode codebook x) (str_list msg))
```

File: contents/huffman_encoding/code/ocaml/huffman.ml [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+  | Some(p1, t1), Some(p2, t2) ->
+      Pq.add q (p1 + p2) (Node(t1, t2));
+      create_huffman_tree q
+  | Some(_, t), None | None, Some(_, t) -> t
+  | None, None                          -> assert false
+
+let rec prefixes_of_tree prefix trees = match trees with
+  | Leaf s -> [(s, prefix)]
+  | Node (t0, t1) -> prefixes_of_tree (prefix ^ "0") t0
+                   @ prefixes_of_tree (prefix ^ "1") t1
```
Comment (ShadowMitia) [inline contents/huffman_encoding/code/ocaml/huffman.ml:84]: You can probably simplify this, and avoid the use of `@` here

**Review by ShadowMitia — CHANGES_REQUESTED** [review]
Globally it looks good and it works. A few things seem to be a bit more complicated than it should be, I've addressed a couple of them.

I would suggest to also create a separate file for the priority queue (and perhaps for some of the helper functions), since it's not the main point of the chapter, it would make the example more readable. But I don't mind it being all in the same file.

File: contents/huffman_encoding/code/ocaml/huffman.ml [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+
+let huffman huffman_tree = prefixes_of_tree "" huffman_tree
+
+(* Helper functions *)
+let char_to_str = Printf.sprintf "%c"
+
+let str_list msg = 
+  List.map char_to_str @@ str_to_charlist msg
```
[the remark this answers is not in the document]
Comment (shubhamkumar13) [reply contents/huffman_encoding/code/ocaml/huffman.ml:92]: The `@@` operator is a function application operator you can alternatively use parens but I usually avoid using parens, it's a habit I picked up using haskell (with `$` operator), if this seems bad I'll remove it

**Comment by shubhamkumar13** [comment]
@ShadowMitia thanks for reviewing this I've left a comment related to the `@@` operator, I think I can remove it since it doesn't matter much either way.

I'll try and replace append function usage, I was also curious how you were able to suggest changes in comments I want to try it, it seems a great way to pinpoint changes.

Thanks again!

**Comment by ShadowMitia** [comment]
@shubhamkumar13 Glad I can help!

For suggestions in github I learned it from [here](https://til.hashrocket.com/posts/szgdi4at59-github-insert-a-suggestion-).

File: contents/huffman_encoding/code/ocaml/huffman.ml [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+let list_to_string lst =
+  List.fold_left (fun acc x -> acc ^ x) "" lst
+
+(* Encoding and decoding functions *)
+let encode codebook x =
+  List.filter (fun (ch, code) -> ch = x) codebook |> fun x ->
+  List.hd x |> snd
+
+let encode_msg codebook msg =
+  List.map (fun x -> encode codebook x) @@ str_list msg |>
```
[the remark this answers is not in the document]
Comment (shubhamkumar13) [reply contents/huffman_encoding/code/ocaml/huffman.ml:99]: Yeah I think you are right here, this seems more readable I'll commit this suggestion

File: contents/huffman_encoding/code/ocaml/huffman.ml (hunk 5 above)
Comment (shubhamkumar13) [reply contents/huffman_encoding/code/ocaml/huffman.ml:80]: Maybe using `List.concat` might be more readable. I'll try that

**Comment by shubhamkumar13** [comment]
@ShadowMitia I've updated the edits let me know if this looks good, Thanks!

**Review by ShadowMitia — APPROVED** [review]
I think this looks good now. Thanks for the work!

**Pull request merged**